src/ua_securechannel.c

```diff
@@ -690,83 +520,269 @@ decryptAddChunk(UA_SecureChannel *channel, UA_MessageType messageType,
 #endif
 
     /* Check (and revolve) the SecurityToken */
-    res = checkSymHeader(channel, tokenId, allowPreviousToken);
+    res = checkSymHeader(channel, tokenId);
     if(res != UA_STATUSCODE_GOOD)
         return res;
 
-    UA_UInt32 requestId = 0;
-    UA_UInt32 sequenceNumber = 0;
-    res = decryptAndVerifyChunk(channel, &sp->symmetricModule.cryptoModule, messageType,
-                                chunk, offset, &requestId, &sequenceNumber);
-    if(res != UA_STATUSCODE_GOOD)
-        return res;
+    /* Decrypt the chunk payload */
+    if(!chunk->decrypted) {
+        res = decryptAndVerifyChunk(channel, &sp->symmetricModule.cryptoModule,
+                                    chunk->messageType, &chunk->bytes, offset);
+        if(res != UA_STATUSCODE_GOOD)
+            return res;
+        chunk->decrypted = true;
+    }
 
     /* Check the sequence number. Skip sequence number checking for fuzzer to
      * improve coverage */
+    UA_SequenceHeader sequenceHeader;
+    res = UA_SequenceHeader_decodeBinary(&chunk->bytes, &offset, &sequenceHeader);
 #ifndef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
-    res = processSequenceNumberSym(channel, sequenceNumber);
-    if(res != UA_STATUSCODE_GOOD)
-        return res;
+    res |= processSequenceNumberSym(channel, sequenceHeader.sequenceNumber);
 #endif
 
-    /* A message consisting of serveral chunks is aborted */
-    if(chunkType == UA_CHUNKTYPE_ABORT) {
-        deleteLatestMessage(channel, requestId);
+    *requestId = sequenceHeader.requestId;
+    return res;
+}
+
+static UA_StatusCode
+assembleProcessMessage(UA_SecureChannel *channel, UA_ChunkQueue *chunks, size_t chunksSize,
+                       UA_UInt32 requestId, UA_MessageType messageType,
+                       void *application, UA_ProcessMessageCallback callback) {
+    /* Only MSG and CLO messages at this point */
+    UA_assert(messageType == UA_MESSAGETYPE_MSG ||
+              messageType == UA_MESSAGETYPE_CLO);
+
+    /* Assemble message */
+    UA_Chunk *chunk;
+    UA_ByteString payload;
+    if(chunksSize > 1) {
+        /* Compute the full message length. Omit the MessageHeader and
```

Comment:
We can avoid this loop by calculating the size when the chunks are extracted from the packet, or when the chunksSize variable is calculated.